feat: experimental static import.meta.env (#12105)
withastro/docs: src/content/docs/en/reference/experimental-flags/static-import-meta-env.mdx

---
title: Experimental static import.meta.env
sidebar:
  label: Static import.meta.env
i18nReady: true
---

import Since from '~/components/Since.astro'

<p>

**Type:** `boolean`<br />
**Default:** `false`<br />
<Since v="5.13.0" />
</p>

Astro allows you to configure a [type-safe schema for your environment variables](/en/guides/environment-variables/#type-safe-environment-variables), and converts variables imported via `astro:env` into the expected type.

> However, Astro by default turns non-public `import.meta.env` values into `process.env` calls during the build, if the environment variable name is present in `process.env`. It also converts your environment variables used through `import.meta.env` in some cases, and this can prevent access to some values such as the strings `"true"` (which is converted to a boolean value), and `"1"` (which is converted to a number).

Reviewer: ```suggestion
However, you can still access your environment variables through `process.env` as well as `import.meta.env` directly if needed. This was the only way to use environment variables in Astro before `astro:env` was added in Astro 5.0, and its handling of `import.meta.env` includes some logic that was intended for earlier versions of Astro that is no longer necessary.

The `experimental.staticImportMetaEnv` flag updates the behavior when accessing `import.meta.env` directly to align with [Vite's handling of environment variables](https://vite.dev/guide/env-and-mode.html#env-variables) and ensures that `import.meta.env` values are always inlined.

Currently, non-public environment variable are replaced by a reference to `process.env`. Additionally, Astro may also convert the value type of your environment variables used through `import.meta.env` which can prevent access to some values such as the strings `"true"` (which is converted to a boolean value), and `"1"` (which is converted to a number).
```

> The `experimental.staticImportMetaEnv` flag disables this behavior, ensuring that `import.meta.env` values are always inlined.
> 
> To enable this feature, add the experimental flag in your Astro config:

Reviewer: ```suggestion
The `experimental.staticImportMetaEnv` flag simplifies Astro's default behavior, making it easier to understand and use. Astro will no longer replace any `import.meta.env` environment variables with a `process.env` call, nor will it coerce values.

To enable this feature, add the experimental flag in your Astro config:
```

```js title="astro.config.mjs" ins={4-6}
import { defineConfig } from "astro/config"

export default defineConfig({
  experimental: {
    staticImportMetaEnv: true,
  }
})
```

## Usage

Enabling this experimental flag will no longer convert string values into booleans or numbers, nor turn `import.meta.env` values into `process.env` calls. This aligns `import.meta.env`'s behavior in Astro with [Vite](https://vite.dev/guide/env-and-mode.html#env-variables).

> In a future major version, Astro will switch to this behavior by default, but you can opt in to the future behavior early using the `experimental.rawEnvValues` flag and if necessary, [updating your project](#updating-your-project) accordingly.

Reviewer 2: renamed:
```suggestion
In a future major version, Astro will switch to this behavior by default, but you can opt in to the future behavior early using the `experimental.staticImportMetaEnv` flag and if necessary, [updating your project](#updating-your-project) accordingly.
```

### Updating your project

If you were relying on coercion, you may need to update your project code to apply it manually:

```ts title="src/components/MyComponent.astro" del={1} ins={2}
const enabled: boolean = import.meta.env.ENABLED
const enabled: boolean = import.meta.env.ENABLED === "true"
```

If you were relying on the transformation into `process.env`, you may need to update your project code to apply it manually:

```ts title="src/components/MyComponent.astro" del={1} ins={2}
const enabled: boolean = import.meta.env.DB_PASSWORD
const enabled: boolean = process.env.DB_PASSWORD
```

> You may need to update types as well:

Reviewer: ```suggestion
You may also need to update types:
```

---

> title: Experimental static import.meta.env
> sidebar:
>   label: Static import.meta.env

Reviewer: ```suggestion
title: Experimental environment variable inlining
sidebar:
  label: Environment variable inlining
```
@florian-lefebvre - Chris and I were talking about having inline code in a page title, and how all the other features were descriptions vs. the name of the flag and we tried to think of a way to *explain* the feature with a descriptive title.

Chris's suggestion was better than my `Experimental not touching your environment variables anymore`! Do you think this accurately describes the feature/content of the page? Is there anything else you'd suggest?

i18nReady: true
---

import Since from '~/components/Since.astro'

<p>

**Type:** `boolean`<br />
**Default:** `false`<br />
<Since v="5.13.0" />
</p>

:::tip[Astro 6.0 preview]
The behavior enabled by this feature will become the default behavior in Astro 6.0.

> You may wish to add this flag whenever it is convenient to do so to start enjoying the benefits sooner, and to avoid the need to update your project code for the next major Astro version.

Reviewer: ```suggestion
You may wish to add this flag whenever it is convenient to do so. You can start enjoying the benefits sooner, and will avoid the need to update your project code for the next major Astro version.
```
